### ISSUE TITLE
White flash only on iOS

### ISSUE THREAD

**Issue by gabriellend** [zoontek/react-native-bootsplash#378]
### Bug summary

Hi!

Not sure this is a bug with this library or just a result of how iOS native code works. A React Native app I work on has been using this library at version 3.2.6 for awhile and it's worked well. The way it's used is a little quirky:

- Load the splash screen (in componentDidMount)
- Hide the splash screen, so we can show a status bar while the app gets any Codepush updates
- I was told there was a white flash happening so to get around it, my predecessors showed the splash screen again briefly after the update and before the landing page.

Now everyone who worked on it before is gone and I'm now attempting to update the library to the latest version. I notice that the `show` method is no more. That would be ok if I could figure out this white flash. It's only happening on iOS (simulator and physical device, iOS 15) and I don't seem to be getting any warnings. Also, if there is no Codepush update and the progress bar is not rendered, there is no white flash. I saw two other issues about white flashes but one was about Android and one was about both Android and iOS and seemed to be related to the splash screen being called before the initial render, but that doesn't seem to be happening here. I also saw that it's related to `presentViewController` being called before the view is correctly presented, which is out of my control (I think?). Does it seem like this is the end of the road, or might there be another reason for the white flash on iOS? If not, I will probably stick with the older version. Thanks very much for any insight!

### Library version

4.3.0

### Environment info

```shell
System:
    OS: macOS 12.2
    CPU: (8) x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
    Memory: 61.63 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.13.1 - ~/.nvm/versions/node/v16.13.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.13.1/bin/yarn
    npm: 8.1.2 - ~/.nvm/versions/node/v16.13.1/bin/npm
    Watchman: 2022.02.14.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.4, iOS 15.4, macOS 12.3, tvOS 15.4, watchOS 8.5
    Android SDK:
      API Levels: 30, 31, 33
      Build Tools: 30.0.2, 30.0.3, 31.0.0, 33.0.0
      System Images: android-30 | Google APIs Intel x86 Atom_64, android-30 | Google Play Intel x86 Atom_64, android-31 | Google APIs Intel x86 Atom_64, android-31 | Google Play Intel x86 Atom_64, android-33 | Google APIs Intel x86 Atom_64, android-33 | Google Play Intel x86 Atom_64
      Android NDK: Not Found
  IDEs:
    Android Studio: 2021.1 AI-211.7628.21.2111.8193401
    Xcode: 13.3.1/13E500a - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.11 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: 0.67.4 => 0.67.4 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

### Steps to reproduce

The project is so large with so many moving parts that I don't think I can provide a reproducible example, apologies. This is more of an intelligence gathering mission. I can provide the part of the code that we use the library in, if that is helpful. You can see we were showing and hiding the splash screen a few times but I've commented everything out except the initial hide.

### Reproducible sample code

```js
async componentDidMount() {
    const { forceUpdate, fadeAnim } = this.state;
    const {
      setApp,
      setUser,
      cleanUser,
      setLocation,
      navigation,
      updatePlatform,
      getCreditCards,
      setUserDocument,
      updateAppVersion,
      getBillingBalance,
    } = this.props;

    if (forceUpdate) {
      await User.logout();
      return NavigationActions.setNewStack(navigation, 'OnboardingLandingRoot');
    }

    updateAppVersion(VersionCheck.getCurrentVersion());

    return RNBootSplash.getVisibilityStatus().then(async (status) => {
      try {
        if (status === 'visible') {
          // Hides the SplashScreen, to show the CodePush download animation.
          RNBootSplash.hide({ fade: true });

          // Comment to not use CodePush.
          await CodePush.sync({
            updateDialog: false,
            installMode: CodePush.InstallMode.IMMEDIATE,
          },
          undefined, // syncStatusChangedCallback
          (progress) => {
            const { receivedBytes, totalBytes } = progress || {};
            this.setState({ progress: receivedBytes / totalBytes });

            // Show SplashScreen to hide white flash.
            // if (receivedBytes / totalBytes >= 1) RNBootSplash.show({ fade: true });
          });
        }

        Animated.timing(fadeAnim, {
          toValue: 0,
          duration: 1000,
          useNativeDriver: true,
        }).start(async () => {
          const userToken = await API.getUserToken();
          if (!userToken) return NavigationActions.setNewStack(navigation, 'OnboardingLandingRoot');

          // The following code only runs if the app has a user token.
          // RNBootSplash.show();

          App.checkDownloadID();

          const res = await Promise.all([
            CodePush.getUpdateMetadata(), // No need to error handle this.
            setUserDocument(),
            getCreditCards(), // If these last two fail, we will handle it at checkout.
            getBillingBalance(),
          ]);

          const metadata = res[0];
          const { label, appVersion } = metadata || { label: '', appVersion: '' };
          updatePlatform({
            build: label.replace('v', ''),
            os: Platform.OS,
            version: appVersion,
          });

          // If setUserDocument fails here, we should take the user back to the login screen.
          const setUserRes = res[1];
          const { payload } = setUserRes || {};
          const { market: marketId } = payload || {};

          const market = await API.getMarket(marketId);
          if (!market) {
            await Promise.all([
              AsyncStorage.removeItem('userToken'),
              AsyncStorage.removeItem('user'),
            ]);

            cleanUser();

            // RNBootSplash.hide({ fade: true });
            return NavigationActions.setNewStack(navigation, 'OnboardingLandingRoot');
          }

          setUser({ previewMenu: false });

          setLocation({ market });

          await Order.updateLatestOrders();

          // Check to see if user has seen notification prompt.
          const APNTokenSeen = await AsyncStorage.getItem('pushEnabled');
          if (APNTokenSeen === null) { await Iterable.updatePushStatus(''); } else await PushNotification.requestPermissions();

          const firstTimeSplashSeen = await AsyncStorage.getItem('firstTimeSplashSeen');
          if (firstTimeSplashSeen === null) setApp({ firstTimeSignUp: true });

          setApp({ comingFromLoadingView: true });
          // RNBootSplash.hide({ fade: true });
          return NavigationActions.setNewStack(navigation, 'Home');
        });
      } catch (err) {
        // eslint-disable-next-line no-console
        console.error(err);
      }
    });
  }
```

**Comment by zoontek** [comment]
H! This is not an issue with the library itself. The v4 doesn't use `UIViewController` anymore and the previous issue mentioning `presentViewController` are not relevant anymore.

This is happening because once CodePush finish downloading the new JS bundle, the JS app has to be restarted. Before, you used `show` to hide this restart since the splash screen is a native view (not handled by the JS). But now that this feature has been removed, you need to find a new trick to hide it.

**Comment by gabriellend** [comment]
@zoontek I super appreciate your fast response and insight, thank you!

**Comment by pke** [comment]
@zoontek why was the ability to toggle the splashscreen removed at some point? It could have helped in that scenario.

**Comment by zoontek** [comment]
@pke Because v4 uses Android 12 SplashScreen API, not a custom API anymore

**Comment by hhemmati81** [comment]
You can remove the white from the rootView and change it to the background color of the splash screen. Go to AppDelegate.m and find the section that is passing the background color.

`if (@available(iOS 13.0, *)) {
    rootView.backgroundColor = [UIColor systemBackgroundColor];
  } else {
    rootView.backgroundColor = [UIColor whiteColor];
  }`

Change the UIColor to your own value like this 
`[[UIColor alloc] initWithRed:0.47 green:0.65 blue:0.87 alpha:1];`

You can convert hex to UIColor value by visiting this page https://www.uicolor.io/

This way you see the bootsplash without the logo for a second before the JS layer loads.

**Comment by gabriellend** [comment]
@hhemmati81 Thanks for your answer and the resources! Sorry, I just saw this right now. That section about background color isn't in my `AppDelegate.m`. I'd like to try this out though, does it matter where I put it? I tried putting it right above what looks like the bootsplash section but I got some build errors about missing brackets.

**Comment by hhemmati81** [comment]
@gabriellend here are my launch options. You can compare it with yours and see what's missing.

```
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  RCTAppSetupPrepareApp(application);

  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];

#if RCT_NEW_ARCH_ENABLED
  _contextContainer = std::make_shared<facebook::react::ContextContainer const>();
  _reactNativeConfig = std::make_shared<facebook::react::EmptyReactNativeConfig const>();
  _contextContainer->insert("ReactNativeConfig", _reactNativeConfig);
  _bridgeAdapter = [[RCTSurfacePresenterBridgeAdapter alloc] initWithBridge:bridge contextContainer:_contextContainer];
  bridge.surfacePresenter = _bridgeAdapter.surfacePresenter;
#endif

  NSDictionary *initProps = [self prepareInitialProps];
  UIView *rootView = RCTAppSetupDefaultRootView(bridge, @"vPresent", initProps);
  if (@available(iOS 13.0, *)) {
    rootView.backgroundColor = [[UIColor alloc] initWithRed:0.47 green:0.65 blue:0.87 alpha:1];
  } else {
    rootView.backgroundColor = [[UIColor alloc] initWithRed:0.47 green:0.65 blue:0.87 alpha:1];
  }
  

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [RNBootSplash initWithStoryboard:@"BootSplash" rootView:rootView];
  return YES;
}
```

**Comment by gabriellend** [comment]
@hhemmati81 Just wanted to circle back, as once I upgraded to React Native 0.70, I had the code you referenced and I was able to make those changes. In 0.71.0, everything changed and I had to figure out a different [way](https://github.com/facebook/react-native/issues/35937#issuecomment-1421608186). But I just wanted to say thank you for you help again and that it worked!
